R/config.R

```diff
@@ -1,45 +1,126 @@
-  if (length(missing) == 0) {
-    return()
-  }
+config_pluck <- function(pkg, path, default = NULL) {
+  check_string(path, allow_empty = FALSE, .internal = TRUE)
 
-  missing_components <- lapply(missing, function(x) c(where, x))
-  msg_flds <- purrr::map_chr(missing_components, paste, collapse = ".")
+  where <- strsplit(path, ".", fixed = TRUE)[[1]]
+  purrr::pluck(pkg$meta, !!!where, .default = default)
+}
 
-  config_abort(
-    pkg, 
-    "Can't find {cli::qty(missing)} component{?s} {.field {msg_flds}}.",
-    call = call
+config_pluck_character <- function(pkg,
+                                   path,
+                                   default = character(),
+                                   call = caller_env()) {
+  x <- config_pluck(pkg, path, default)
+  config_check_character(
+    x,
+    error_path = path,
+    error_pkg = pkg,
+    error_call = call
   )
 }
 
-yaml_character <- function(pkg, where) {
-  x <- purrr::pluck(pkg$meta, !!!where)
+config_pluck_string <- function(pkg,
+                                path,
+                                default = "",
+                                call = caller_env()) {
+  x <- config_pluck(pkg, path, default)
+  config_check_string(
+    x,
+    error_path = path,
+    error_pkg = pkg,
+    error_call = call
+  )
+}
+
+# checks ---------------------------------------------------------------------
 
-  if (identical(x, list()) || is.null(x)) {
+config_check_character <- function(x,
+                                   error_pkg,
+                                   error_path,
+                                   error_call = caller_env()) {
+  if (is.character(x)) {
+    x
+  } else if (identical(x, list())) {
     character()
-  } else if (is.character(x)) {
+  } else {
+    config_abort_type(
+      must_be = "a character vector",
+      not = x,
+      error_pkg = error_pkg,
+      error_path = error_path,
+      error_call = error_call
+    )
+  }
+}
+
+config_check_string <- function(x,
+                                error_pkg,
+                                error_path,
+                                error_call = caller_env()) {
+
+  if (is_string(x)) {
     x
   } else {
-    path <- paste0(where, collapse = ".")
+    config_abort_type(
+      must_be = "a string",
+      not = x,
+      error_pkg = error_pkg,
+      error_path = error_path,
+      error_call = error_call
+    )
+  }
+}
+
+config_abort_type <- function(must_be, not, error_pkg, error_path, error_call) {
+  not_str <- obj_type_friendly(not)
+  config_abort(
+    error_pkg,
+    "{.field {error_path}} must be {must_be}, not {not_str}.",
+    call = error_call
+  )
+
+}
+
+config_check_list <- function(x,
+                              names = NULL,
+                              error_pkg,
+                              error_path,
+                              error_call = caller_env()) {
+  if (is_list(x)) {
+    if (!is.null(names) && !all(has_name(x, names))) {
```

Comment:
based on my confusion in the tests before I read the error snapshots:
```suggestion
    some_required_components_absent <- (!is.null(names) && !all(has_name(x, names)))
    if  some_required_components_absent {
```

`names` is maybe not specific enough, could it be "required" or so.

Reply:
Maybe `has_names`?

Reply:
has_required? names is a bit vague

Reply:
I tried `has_names` and I think it's pretty clear when read in context.